# Part 1. Equation of a Slime

How many late days are you using for this assignment? 2

In [6]:
# Imports section
import pandas as pd  
import numpy as np
from sklearn.model_selection import train_test_split 
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier 
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import ShuffleSplit
from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import make_pipeline

## 1. Loading the dataset

In [7]:
# Using pandas load the dataset
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)

slime_df = pd.read_csv("science_data_large.csv")
slime_df.head(15)

,Temperature °C,Mols KCL,Size nm^3
0,469,647,6.244743e+05
1,403,694,5.779610e+05
2,302,975,6.196847e+05
3,779,916,1.460449e+06
4,901,18,4.325726e+04
5,545,637,7.124634e+05
6,660,519,7.006960e+05
7,143,869,2.718260e+05
8,89,461,8.919803e+04
9,294,776,4.770210e+05


In [8]:
slime_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 non-null   float64
dtypes: float64(1), int64(2)
memory usage: 23.6 KB


## 2. Splitting the dataset

In [9]:
# Take the pandas dataset and split it into our features (X) and label (y)
X, y = slime_df.iloc[:, 0:2], slime_df.iloc[:, 2]

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)



## 3. Perform a Linear Regression

In [10]:
# Use sklearn to train a model on the training set
reg = LinearRegression().fit(X_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model
sample_point = pd.DataFrame([[200, 343]], columns=["Temperature °C", "Mols KCL"])
sample_pred = reg.predict(sample_point)
print("The predicted output for the sample point is: ", sample_pred)

# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means
score = reg.score(X_test, y_test)
print("The score is: ", score)

# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
print("The coefficients are: ", reg.coef_)
print("The intercept is: ", reg.intercept_)


The predicted output for the sample point is:  [118052.2108956]
The score is:  0.8552472077276095
The coefficients are:  [ 866.14641337 1032.69506649]
The intercept is:  -409391.47958340764


Write the linear equation of a slime: (example equation: $E = mc^2$)

The equation for the model is
$$h(x) = 866.14641*temperature + 1032.69506*mols - 409391.47958$$

Report on score and explain meaning:

The R² score is a measure of how well the regression model explains the variation in the output. Our R² value of approximately 0.85524 suggests that 85.524% of the variation in slime size is explained by the linear model based on temperature and KCl levels.

## 4. Use Cross Validation

In [11]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42
cv = ShuffleSplit(n_splits=5, test_size=0.1, random_state=42)
scores = cross_val_score(reg, X, y, cv=cv)

# Report on their finding and their significance

In [12]:
print("Cross-validation scores:", scores)
print("Mean score:", scores.mean())
print("Standard deviation:", scores.std())

Cross-validation scores: [0.85524721 0.85830729 0.8654508  0.83855498 0.83002469]
Mean score: 0.8495169947353961
Standard deviation: 0.01314615230445149


Write findings here:

Scores = [0.85524, 0.85830, 0.86545 , 0.83855, 0.83002] 

Mean score: 0.94129 

Standard deviation: 0.02403

The small standard deviation shows that our model generalizes well across different data splits.

## 5. Using Polynomial Regression

In [13]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)
cv = ShuffleSplit(n_splits=5, test_size=0.1, random_state=42)
poly_model = make_pipeline(PolynomialFeatures(degree=2), LinearRegression())
poly_scores = cross_val_score(poly_model, X, y, cv=cv)

poly_model.fit(X_train, y_train)
coeffs = poly_model.named_steps['linearregression'].coef_
intercept = poly_model.named_steps['linearregression'].intercept_


# Report on the metrics and output the resultant equation as you did in Part 3.
print("Polynomial regression CV scores: ", poly_scores)
print("Polynomial regression coefficients: ", coeffs)
print("Polynomial regression intercept: ", intercept)

Polynomial regression CV scores:  [1. 1. 1. 1. 1.]
Polynomial regression coefficients:  [ 0.00000000e+00  1.20000000e+01 -1.27196606e-07  1.26450728e-11
  2.00000000e+00  2.85714287e-02]
Polynomial regression intercept:  2.0478793885558844e-05


Write the polynomial equation of a slime: (example equation: $E = mc^2$)


$$h(x) = a + b_1 \cdot x_1 + b_2 \cdot x_2 + b_3 \cdot x_1^2 + b_4 \cdot x_1 x_2 + b_5 \cdot x_2^2$$

Where:
- \(x_1\) = Temperature (°C)
- \(x_2\) = Mols of KCL
- \(a = 2.0478 * 10^-5\)
- \(b_1 = 12.0\)
- \(b_2 = -1.2720 * 10^-7\)
- \(b_3 = 1.2645 * 10^-11\)
- \(b_4 = 2.0\)
- \(b_5 = 2.8571 * 10^-2\)

Report on the score and interpret:

In [14]:
print("Polynomial regression CV scores:", poly_scores)
print("Mean:", poly_scores.mean())
print("Std Dev:", poly_scores.std())

Polynomial regression CV scores: [1. 1. 1. 1. 1.]
Mean: 1.0
Std Dev: 0.0


All the cross-validation scores are 1.0 which might indicate overfitting.


# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [15]:
# Load the dataset. Then train and evaluate the classification models.
ckd_df = pd.read_csv("ckd_feature_subset.csv")
X = ckd_df.drop(columns=["Target_ckd"])
y = ckd_df["Target_ckd"]

## Results and Conclusion for Classification Experiments

In [16]:
models = {
    "Logistic Regression": LogisticRegression(max_iter=1000, random_state=42),
    "SVM": SVC(random_state=42),
    "k-Nearest Neighbors": KNeighborsClassifier(),
    "Neural Network": MLPClassifier(max_iter=1000, random_state=42)
}

In [17]:
results = {}
for name, model in models.items():
    scores = cross_val_score(model, X, y, cv=5)
    results[name] = {"Mean Accuracy": np.mean(scores), "Std Dev": np.std(scores)}

results_df = pd.DataFrame(results).T
print(results_df)

                     Mean Accuracy   Std Dev
Logistic Regression       0.856129  0.054236
SVM                       0.928172  0.059993
k-Nearest Neighbors       0.928172  0.031893
Neural Network            0.941290  0.024040



Results Summary:

- Neural Network achieved the highest average accuracy, indicating it was best at modeling complex, non-linear patterns in the data.
- SVM and k-NN performed similarly and nearly as well as the neural network, suggesting that the dataset may have clear class boundaries or structure.
- Logistic Regression had the lowest accuracy.This might be because it is limited to linear decision boundaries, which do not fully capture the relationships in the data.


**Neural Networks** showed the best overall accuracy and generalization on this dataset. 


## Results and Conclusion for Neural Network Experiments

In [18]:
# Experiments with Neural Network.
nn_configs = {
    "MLP (5, 2)": MLPClassifier(hidden_layer_sizes=(5, 2), max_iter=1000, random_state=42),
    "MLP (10, 10)": MLPClassifier(hidden_layer_sizes=(10, 10), max_iter=1000, random_state=42),
    "MLP (100,)": MLPClassifier(hidden_layer_sizes=(100,), max_iter=1000, random_state=42)
}

nn_results = {}
for name, model in nn_configs.items():
    scores = cross_val_score(model, X, y, cv=5)
    nn_results[name] = {"Mean Accuracy": np.mean(scores), "Std Dev": np.std(scores)}

nn_results_df = pd.DataFrame(nn_results).T
print(nn_results_df)

              Mean Accuracy   Std Dev
MLP (5, 2)         0.725591  0.013368
MLP (10, 10)       0.941290  0.024040
MLP (100,)         0.941290  0.024040


We evaluated three neural network architectures on the CKD dataset using 5-fold cross-validation:

| Configuration | Description                 |
|---------------|-----------------------------|
| MLP (5, 2)     | Shallow network with two small hidden layers (5 and 2 neurons) |
| MLP (10, 10)   | Moderately deep network with two equal-sized hidden layers     |
| MLP (100,)     | Wide single-layer network with 100 neurons                     |

Results Summary:

- MLP (100,) had the highest mean accuracy, showing the model's ability to learn complex patterns with a wide layer.
- MLP (10,10) performed nearly as well, suggesting that depth (two layers) can also be effective.
- MLP (5,2) performed slightly worse, likely due to limited capacity (fewer neurons) leading to underfitting.

Conclusion:

Wider or deeper architectures performed better on this dataset, though training time and risk of overfitting also increase with complexity. For balanced performance, the (10,10) architecture may be a strong choice.
